Core 11737 keyboard layout hotkey

thc13 opened the pull request:
**!!!Please keep commit history at merge!!!**

## Purpose

Hotkey for Language change is buggy:
Following points are not equal to winxp (win2k3):
1. hotkey behavior, is not independent from press/release order of the keys
2. tray icon reflects language selection in windows and not layout, but until only one layout is associated with the language it has the same effect, but in ReactOS it is actually the layout
3. Fix check mark handling in popup menu of tray icon, it is not updated by hotkey
4. Find and fix reason, why it only works if language is install at setup time, but not later.
4.1 Actual guess, not all registry keys are updated, there is a difference to winxp behavior and also a mismatch between installing additional languages at install time or later in running system   
5. keyboard layout (language) change is global atm, but in winxp it is process/window dependent
6. support multiple keyboard layouts per language

JIRA issue: [CORE-11737](https://jira.reactos.org/browse/CORE-11737)

## Proposed changes

Actual PR addresses topics 1-3

## TODO

Topic 4, will be addressed in a separate PR. So no need to wait here. If fix is available before merge, i will add commit here. But it should be for sure in context of [CORE-11737](https://jira.reactos.org/browse/CORE-11737)

Topic 5 + 6 are out of scope of this PR and imo also of [CORE-11737](https://jira.reactos.org/browse/CORE-11737)

binarymaster commented:
Suggestions to rename commits properly:
```diff
-[hotkey] Language change registry key values and read
+[SETUPLIB][NTUSER] Fix keyboard toggle registry key and its default values
```

```diff
-[hotkey] Make language change sequence independent
+[NTUSER] Make the language change sequence-independent
```

```diff
-[kbswitch]: Keep check mark for language selection in sync
+[KBSWITCH] Keep check mark for language selection in sync
```

Prefix in each commit message stands for the module(s) you change.

Also you need to add "CORE-11737" as a last line to each commit message, so commits are associated with Jira ticket. Variations are possible, but better follow this as a standard.

See this commit message for example: 2538583dbe13560676ac9bb3dc0911b8a070c1a1.

thc13 commented:
## TODO
Topic 4, is addressed with improved registry key handling and attached workaround patch in [CORE-11737](https://jira.reactos.org/browse/CORE-11737), for more details please look into ticket.

Topic 5 + 6 are out of scope of this PR and imo also of CORE-11737

HBelusca commented on base/applications/kbswitch/kbswitch.c:
```diff
@@ -436,6 +436,7 @@ WndProc(HWND hwnd, UINT Message, WPARAM wParam, LPARAM lParam)
 
         case WM_LANG_CHANGED:
         {
+            ulCurrentLayoutNum=GetNextLayout();
```
```suggestion
            ulCurrentLayoutNum = GetNextLayout();
```

thc13 replied:
Done

HBelusca commented on dll/cpl/input/input_list.c:
```diff
@@ -323,14 +323,56 @@ InputList_AddInputMethodToUserRegistry(DWORD dwIndex, INPUT_LIST_NODE *pNode)
+     * First time more then 1 input language exists switching hotkeys needed
+     * set default values for language/layout switching hotkeys
+     */
+    if (2 == dwIndex) {
```
Please let's keep existing (and if possible, sane [code style](https://reactos.org/wiki/Coding_Style) ):
```suggestion
    if (dwIndex == 2)
    {
```

thc13 replied:
Done

HBelusca commented on dll/cpl/input/input_list.c:
```diff
@@ -323,14 +323,56 @@ InputList_AddInputMethodToUserRegistry(DWORD dwIndex, INPUT_LIST_NODE *pNode)
+     */
+    if (2 == dwIndex) {
+        if (RegOpenKeyExW(HKEY_CURRENT_USER,
+                      L"Keyboard Layout\\Toggle",
```
Please fix indentation (here and in the other functions below).

thc13 replied:
done

HBelusca commented on dll/cpl/input/input_list.c:
```diff
@@ -323,14 +323,56 @@ InputList_AddInputMethodToUserRegistry(DWORD dwIndex, INPUT_LIST_NODE *pNode)
     }
 
+    /* 
+     * First time more then 1 input language exists switching hotkeys needed
+     * set default values for language/layout switching hotkeys
```
What do you exactly mean in this comment?

thc13 replied:
Adjusted hopefully it is clear now

HBelusca commented on dll/cpl/input/input_list.c:
```diff
@@ -323,14 +323,56 @@ InputList_AddInputMethodToUserRegistry(DWORD dwIndex, INPUT_LIST_NODE *pNode)
+
+            RegCloseKey(hKey);
+        }
+    }
```
If this code block is part of the workaround you mentioned in the JIRA report, please mention that explicitly in the code!

thc13 replied:
No it is not, workaround is in a separate patch only attached to JIRA

HBelusca commented on win32ss/user/ntuser/msgqueue.c:
```diff
@@ -1771,6 +1771,8 @@ BOOL co_IntProcessKeyboardMessage(MSG* Msg, BOOL* RemoveMessages)
     WPARAM wParam = Msg->wParam;
     PTHREADINFO pti = PsGetCurrentThreadWin32Thread();
+    UINT language_hotkey;
+    BOOL change_language = FALSE;
```
code style is: `LanguageHotkey` and `ChangeLanguage` (or `bChangeLanguage`)

thc13 replied:
Done

HBelusca commented on win32ss/user/ntuser/msgqueue.c:
```diff
@@ -1870,46 +1871,74 @@ BOOL co_IntProcessKeyboardMessage(MSG* Msg, BOOL* RemoveMessages)
-                                     WM_INPUTLANGCHANGEREQUEST,
-                                     wParamILR,
-                                    (LPARAM)pkl->hkl );
-
```
Please reintroduce the new line after this `UserPostMessage( WM_INPUTLANGCHANGEREQUEST )` call.

thc13 replied:
Adjusted other places too

thc13 commented:
Adjusted code format add comments. Changes are force pushed, because i would like to keep the commit history. Commit history for commits with, review adjustments make no sense imo.

JoachimHenze commented on dll/cpl/input/input_list.c:
```diff
@@ -323,14 +323,57 @@ InputList_AddInputMethodToUserRegistry(DWORD dwIndex, INPUT_LIST_NODE *pNode)
 
+    /* 
+     * Adding new language to system, if it is the second language, switching
+     * make sense. Activate it hotkeys therefor. Until only one language/layout
```
it hotkeys -> it**s** hotkeys?

thc13 replied:
Adjusted and rebased to master

HBelusca commented on win32ss/user/ntuser/msgqueue.c:
```diff
@@ -1871,45 +1872,75 @@ BOOL co_IntProcessKeyboardMessage(MSG* Msg, BOOL* RemoveMessages)
             {
-                WPARAM wParamILR;
-                PKL pkl = pti->KeyboardLayout;
+               if ( Msg->wParam == VK_SHIFT && !(IS_KEY_DOWN(gafAsyncKeyState, VK_SHIFT))) {
```
```suggestion
               if (Msg->wParam == VK_SHIFT && !(IS_KEY_DOWN(gafAsyncKeyState, VK_SHIFT)))
               {
```

HBelusca commented on win32ss/user/ntuser/msgqueue.c:
```diff
@@ -1871,45 +1872,75 @@ BOOL co_IntProcessKeyboardMessage(MSG* Msg, BOOL* RemoveMessages)
+                           && !(IS_KEY_DOWN(gafAsyncKeyState, VK_CONTROL)) )
+                  {
+                     if ( IS_KEY_DOWN(gafAsyncKeyState, VK_LSHIFT) )
+                        gLanguageToggleKeyState = INPUTLANGCHANGE_FORWARD;  // Left Alt - Left Shift, Next
```
Aren't you looking at the Left/Right control key (+ the left/right shift) instead? This seems to change the keyboard layout only, within a fixed language.

thc13 replied:
Language in most cases also changes the layout too. And hotkey is configurable. Actually ROS does not support layout change consistently, there are only parts implemented yet. So this is the nearest for JIRA ticket context. There are future todos but not in this context.

thc13 commented on base/applications/kbswitch/kbswitch.c:
```diff
@@ -436,6 +436,7 @@ WndProc(HWND hwnd, UINT Message, WPARAM wParam, LPARAM lParam)
 
         case WM_LANG_CHANGED:
         {
+            ulCurrentLayoutNum = GetNextLayout();
```
My test show that it is only needed here, but i may missed something. But for sure it is not needed on every call. Because e.g., if you change it with mouse and gui, the correct item is selected already.

thc13 commented on dll/cpl/input/input_list.c:
```diff
@@ -323,14 +323,57 @@ InputList_AddInputMethodToUserRegistry(DWORD dwIndex, INPUT_LIST_NODE *pNode)
+     */
+    if ( dwIndex == 2 ) {
+        if (RegOpenKeyExW(HKEY_CURRENT_USER,
+                            L"Keyboard Layout\\Toggle",
+                            0,
+                            KEY_SET_VALUE,
+                            &hKey) == ERROR_SUCCESS)
```
This is how it is done in old code parts, so i want to keep this or should the rest also be adjusted? I would prefer to only touch code if functional change is needed. If you request that, i will do it.

HBelusca replied:
other similar code had it correct.

thc13 replied:
Done, used your suggestion from below.

thc13 commented on win32ss/user/ntuser/msgqueue.c:
```diff
@@ -1871,45 +1872,75 @@ BOOL co_IntProcessKeyboardMessage(MSG* Msg, BOOL* RemoveMessages)
 
-        if ( gdwLanguageToggleKey < 3 &&
-             IS_KEY_DOWN(gafAsyncKeyState, gdwLanguageToggleKey == 1 ? VK_LMENU : VK_CONTROL) )
+        if ( gdwLanguageToggleKey < 3 )
```
Done

thc13 commented on win32ss/user/ntuser/msgqueue.c:
```diff
@@ -1871,45 +1872,75 @@ BOOL co_IntProcessKeyboardMessage(MSG* Msg, BOOL* RemoveMessages)
-            if ( Msg->wParam == VK_SHIFT && !(IS_KEY_DOWN(gafAsyncKeyState, VK_SHIFT)))
+            LanguageHotkey = gdwLanguageToggleKey == 1 ? VK_LMENU : VK_CONTROL;
+            if( IS_KEY_DOWN(gafAsyncKeyState, LanguageHotkey) ||
+               IS_KEY_DOWN(gafAsyncKeyState, VK_SHIFT) )
```
Done

thc13 commented on win32ss/user/ntuser/msgqueue.c:
```diff
@@ -1871,45 +1872,75 @@ BOOL co_IntProcessKeyboardMessage(MSG* Msg, BOOL* RemoveMessages)
+               if ( Msg->wParam == VK_SHIFT && !(IS_KEY_DOWN(gafAsyncKeyState, VK_SHIFT))) {
+                  bChangeLanguage = TRUE;
+               }
+               else {
```
Done

thc13 commented on dll/cpl/input/input_list.c:
```diff
@@ -323,14 +323,57 @@ InputList_AddInputMethodToUserRegistry(DWORD dwIndex, INPUT_LIST_NODE *pNode)
+                             0,
+                             REG_SZ,
+                             (LPBYTE)szBuffer,
+                             (wcslen(szBuffer) + 1) * sizeof(WCHAR));
```
Done

thc13 commented on dll/cpl/input/input_list.c:
```diff
@@ -323,14 +323,57 @@ InputList_AddInputMethodToUserRegistry(DWORD dwIndex, INPUT_LIST_NODE *pNode)
+     * make sense. Activate hotkeys therefor. Until only one language/layout
+     * is available, switching should be disabled in registry.
+     */
+    if ( dwIndex == 2 ) {
```
Done

thc13 commented:
## New
- Additional commit introduces correct registry setting, when deleting input languages and it is the last one in system.
- Fixed typo and style
- Rebased to master head

## Out of scope topics
Topic 5 + 6, imo also not for [CORE-11737](https://jira.reactos.org/browse/CORE-11737)
5. keyboard layout (language) change is global atm, but in winxp it is process/window dependent
6. support multiple keyboard layouts per language

### Additional out of scope topic:
7. avoid need for reboot, actual test show a bug which is hidden by reboot
- If reboot is canceled hotkeys work fine and language/layout switch takes effect, but a **second tray icon is printed**.

katahiromz commented on dll/cpl/input/input_list.c:
```diff
@@ -323,14 +323,56 @@ InputList_AddInputMethodToUserRegistry(DWORD dwIndex, INPUT_LIST_NODE *pNode)
+                            &hKey) == ERROR_SUCCESS)
+        {
+            /* 'Hotkey' and 'Layout Hotkey' have always the same value */
+            RegSetValueExW(hKey, L"Hotkey", 0, REG_SZ, (LPBYTE)L"1", (sizeof(L"1")));
+            RegSetValueExW(hKey, L"Language Hotkey", 0, REG_SZ, (LPBYTE)L"1", (sizeof(L"1")));
+
+            RegSetValueExW(hKey, L"Layout Hotkey", 0, REG_SZ, (LPBYTE)L"2", (sizeof(L"2")));
```
```suggestion
            RegSetValueExW(hKey, L"Hotkey", 0, REG_SZ, (LPBYTE)L"1", sizeof(L"1"));
            RegSetValueExW(hKey, L"Language Hotkey", 0, REG_SZ, (LPBYTE)L"1", sizeof(L"1"));

            RegSetValueExW(hKey, L"Layout Hotkey", 0, REG_SZ, (LPBYTE)L"2", sizeof(L"2"));
```

katahiromz commented on dll/cpl/input/input_list.c:
```diff
@@ -323,14 +323,56 @@ InputList_AddInputMethodToUserRegistry(DWORD dwIndex, INPUT_LIST_NODE *pNode)
+                            &hKey) == ERROR_SUCCESS)
+        {
+            /* 'Hotkey' and 'Layout Hotkey' have always the same value */
+            RegSetValueExW(hKey, L"Hotkey", 0, REG_SZ, (LPBYTE)L"3", (sizeof(L"3")));
+            RegSetValueExW(hKey, L"Language Hotkey", 0, REG_SZ, (LPBYTE)L"3", (sizeof(L"3")));
+
+            RegSetValueExW(hKey, L"Layout Hotkey", 0, REG_SZ, (LPBYTE)L"3", (sizeof(L"3")));
```
```suggestion
            RegSetValueExW(hKey, L"Hotkey", 0, REG_SZ, (LPBYTE)L"3", sizeof(L"3"));
            RegSetValueExW(hKey, L"Language Hotkey", 0, REG_SZ, (LPBYTE)L"3", sizeof(L"3"));

            RegSetValueExW(hKey, L"Layout Hotkey", 0, REG_SZ, (LPBYTE)L"3", sizeof(L"3"));
```

thc13 replied:
Done

katahiromz commented on win32ss/user/ntuser/msgqueue.c:
```diff
@@ -1871,45 +1872,77 @@ BOOL co_IntProcessKeyboardMessage(MSG* Msg, BOOL* RemoveMessages)
+        if (gdwLanguageToggleKey < 3)
         {
-            if ( Msg->wParam == VK_SHIFT && !(IS_KEY_DOWN(gafAsyncKeyState, VK_SHIFT)))
+            LanguageHotkey = gdwLanguageToggleKey == 1 ? VK_LMENU : VK_CONTROL;
```
```suggestion
            LanguageHotkey = ((gdwLanguageToggleKey == 1) ? VK_LMENU : VK_CONTROL);
```

thc13 replied:
Done

katahiromz commented on win32ss/user/ntuser/msgqueue.c:
```diff
@@ -1871,45 +1872,77 @@ BOOL co_IntProcessKeyboardMessage(MSG* Msg, BOOL* RemoveMessages)
         {
-            if ( Msg->wParam == VK_SHIFT && !(IS_KEY_DOWN(gafAsyncKeyState, VK_SHIFT)))
+            LanguageHotkey = gdwLanguageToggleKey == 1 ? VK_LMENU : VK_CONTROL;
+            if(IS_KEY_DOWN(gafAsyncKeyState, LanguageHotkey) ||
+               IS_KEY_DOWN(gafAsyncKeyState, VK_SHIFT))
```
```suggestion
            if (IS_KEY_DOWN(gafAsyncKeyState, LanguageHotkey) ||
                IS_KEY_DOWN(gafAsyncKeyState, VK_SHIFT))
```

thc13 replied:
Done

katahiromz commented on win32ss/user/ntuser/msgqueue.c:
```diff
@@ -1871,45 +1872,77 @@ BOOL co_IntProcessKeyboardMessage(MSG* Msg, BOOL* RemoveMessages)
             {
-                WPARAM wParamILR;
-                PKL pkl = pti->KeyboardLayout;
+               if (Msg->wParam == VK_SHIFT && !(IS_KEY_DOWN(gafAsyncKeyState, VK_SHIFT)))
```
```suggestion
               if (Msg->wParam == VK_SHIFT && !IS_KEY_DOWN(gafAsyncKeyState, VK_SHIFT))
```

thc13 replied:
I would prefer to keep it as is, but if needed i will change it.

katahiromz commented on win32ss/user/ntuser/msgqueue.c:
```diff
@@ -1871,45 +1872,77 @@ BOOL co_IntProcessKeyboardMessage(MSG* Msg, BOOL* RemoveMessages)
+               }
+               else
+               {
+                  if ( wParam == VK_LMENU && wParam == LanguageHotkey
+                       && !(IS_KEY_DOWN(gafAsyncKeyState, VK_LMENU)) )
```
```suggestion
                  if (wParam == VK_LMENU && wParam == LanguageHotkey &&
                      !IS_KEY_DOWN(gafAsyncKeyState, VK_LMENU))
```

thc13 replied:
Done.

katahiromz commented on win32ss/user/ntuser/msgqueue.c:
```diff
@@ -1871,45 +1872,77 @@ BOOL co_IntProcessKeyboardMessage(MSG* Msg, BOOL* RemoveMessages)
+
+                     bChangeLanguage = TRUE;
+                  }
+                  else if ( Msg->wParam == VK_CONTROL && Msg->wParam == LanguageHotkey
+                           && !(IS_KEY_DOWN(gafAsyncKeyState, VK_CONTROL)) )
```
```suggestion
                  else if (Msg->wParam == VK_CONTROL && Msg->wParam == LanguageHotkey &&
                           !IS_KEY_DOWN(gafAsyncKeyState, VK_CONTROL))
```

thc13 replied:
Done

thc13 commented:
Implemented review findings, please reopen. Or should i open a new PR? If these changes are unwanted, please comment and i leave this untouched in the future.

binarymaster commented:
@thc13 I'm unable to reopen it:

![image](https://user-images.githubusercontent.com/578406/174504424-97b455b7-b582-4379-a263-f5fc87f9b621.png)

HBelusca commented:
> @thc13 I'm unable to reopen it:

It seems to be a "known" issue without any current solution:
https://github.com/isaacs/github/issues/361
https://stackoverflow.com/a/50091366
so I think, the best course of action is to recreate a PR.
